File: 2021/day02/python/part2.py
from dataclasses import dataclass
from typing import List, Tuple


@dataclass
class Position:
    hpos: int
    depth: int
    aim: int


def parse_line(line: str) -> Tuple[str, int]:
    parts = line.partition(" ")
    dir = parts[0]
    amt = int(parts[2])
    return dir, amt


def compute_answer(lines: List[str]) -> int:
    pos = Position(0, 0, 0)
    for line in lines:
        dir, amt = parse_line(line)

        if dir == "forward":
            pos.hpos += amt
            pos.depth += pos.aim * amt
        elif dir == "down":
            pos.aim += amt
        elif dir == "up":
            pos.aim -= amt
        else:
            raise Exception(f"Unknown direction: {dir}")

        assert pos.depth >= 0

    return pos.hpos * pos.depth

File: 2021/day02/rust/src/part2.rs
struct Position {
    hpos: i32,
    depth: i32,
    aim: i32,
}

pub fn parse_line(line: &str) -> (&str, i32) {
    let tokens: Vec<&str> = line.split(" ").collect();
    let dir = tokens[0];
    let amt = tokens[1].parse().unwrap();
    return (dir, amt);
}

pub fn compute_answer(lines: &Vec<&str>) -> i32 {
    let mut pos = Position {
        hpos: 0,
        depth: 0,
        aim: 0,
    };
    for index in 0..lines.len() {
        let (dir, amt) = parse_line(lines[index]);
        match dir {
            "forward" => {
                pos.hpos += amt;
                pos.depth += pos.aim * amt
            }
            "down" => pos.aim += amt,
            "up" => pos.aim -= amt,
            _ => panic!("Unknown direction: {}", dir),
        }
    }
    if pos.depth < 0 {
        panic!("Invalid depth: {}", pos.depth);
    }
    return pos.hpos * pos.depth;
}
